### PR TITLE
respect CUDA_HOST_COMPILER when detecting CUDA arch

### PR DESCRIPTION
**Pull request by cyyever**
```
force_builders=Custom
buildworker:Custom=linux-4
build_image:Custom=ubuntu-cuda:18.04
```

### CODE REVIEW

File: cmake/OpenCVDetectCUDA.cmake [hunk 1]
```diff
@@ -86,12 +86,17 @@ if(CUDA_FOUND)
     unset(CUDA_ARCH_PTX CACHE)
   endif()
 
-  SET(DETECT_ARCHS_COMMAND "${CUDA_NVCC_EXECUTABLE}" ${CUDA_NVCC_FLAGS} "${OpenCV_SOURCE_DIR}/cmake/checks/OpenCVDetectCudaArch.cu" "--run")
-  if(WIN32 AND CMAKE_LINKER) #Workaround for VS cl.exe not being in the env. path
-    get_filename_component(host_compiler_bindir ${CMAKE_LINKER} DIRECTORY)
-    SET(DETECT_ARCHS_COMMAND ${DETECT_ARCHS_COMMAND} "-ccbin" "${host_compiler_bindir}")
+  if(CUDA_HOST_COMPILER)
+    LIST(APPEND CUDA_NVCC_FLAGS -ccbin ${CUDA_HOST_COMPILER})
```
Comment (asmorkalov) [inline cmake/OpenCVDetectCUDA.cmake:90]: `CUDA_HOST_COMPILER` is path to executable, but not directory as it's done in else branch.

Comment (cyyever) [reply cmake/OpenCVDetectCUDA.cmake:90]: I wanted to avoid specifying -ccbin twice.

Comment (cyyever) [reply cmake/OpenCVDetectCUDA.cmake:90]: > `CUDA_HOST_COMPILER` is path to executable, but not directory as it's done in else branch.

If the user specifies CUDA_HOST_COMPILER, we wouldn't need to help him detect cl.exe

> `CUDA_HOST_COMPILER` is path to executable, but not directory as it's done in else branch.

4.2.1.11. --compiler-bindir directory (-ccbin)

Specify the directory in which the compiler executable resides.

**The host compiler executable name can be also specified to ensure that the correct host compiler is selected**. In addition, driver prefix options (--input-drive-prefix, --dependency-drive-prefix, or --drive-prefix) may need to be specified, if nvcc is executed in a Cygwin shell or a MinGW shell on Windows.

**Comment by tomoaki0705** [comment]
This PR brakes the build on Jetson TX1 and TX2

- cmake
```
-- General configuration for OpenCV 4.4.0-pre =====================================
--   Version control:               4.3.0-464-g5ed6546
-- 
--   Extra modules:
--     Location (extra):            /opencv_contrib/modules
--     Version control (extra):     4.3.0-69-g0f56e6d
 :
-- CUDA detected: 8.0
-- Automatic detection of CUDA generation failed. Going to build for all known architectures.
-- CUDA NVCC target flags: -ccbin;/usr/bin/cc;-gencode;arch=compute_53,code=sm_53;-gencode;arch=compute_62,code=sm_62;-gencode;arch=compute_72,code=sm_72;-D_FORCE_INLINES
```
 - It usually detects the native CC (i.e. sm53 for TX1, sm62 for TX2)
   - Then, sm72 is not in the list of CUDA 8.0. (i.e. The fall safe path is not safe at all)

- make
````
nvcc fatal   : Unsupported gpu architecture 'compute_72'
CMake Error at cuda_compile_1_generated_gpu_mat.cu.o.Release.cmake:219 (message):
  Error generating
  /opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o
````

**Comment by NHarishGit** [comment]
This change has broken CUDA builds.

**Comment by tomoaki0705** [comment]
@NHarishGit , I appreciate your comment.
It would be a great help if you could share what "CUDA builds" you are talking about. like [this](https://github.com/opencv/opencv/issues/17598#issuecomment-647125305) 

>     * Jetson TX2 (TX1)
>     * Ubuntu 16.04 (Aarch64)
>     * CUDA 8.0
>     * GCC 5.4.0
>     * CMake 3.15.1

OS, architecture, CUDA version, compiler version, cmake version.
That helps the discussion.
Thanks

**Comment by NHarishGit** [comment]
> @NHarishGit , I appreciate your comment.
> It would be a great help if you could share what "CUDA builds" you are talking about. like [this](https://github.com/opencv/opencv/issues/17598#issuecomment-647125305)
> 
> > ```
> > * Jetson TX2 (TX1)
> > * Ubuntu 16.04 (Aarch64)
> > * CUDA 8.0
> > * GCC 5.4.0
> > * CMake 3.15.1
> > ```
> 
> OS, architecture, CUDA version, compiler version, cmake version.
> That helps the discussion.
> Thanks

OS: WIN 10 x64 2004 build
VS 2019 community v 16.6.2
CUDA v 10.2
CMAKE v 1.17.3
GSTREAMER v 1.16.2
MKL, TBB 2020 versions
CUDA DNN v 7.6.5

**Comment by tomoaki0705** [comment]
@cyyever any comments ?
https://github.com/opencv/opencv/issues/17598#issuecomment-647125305

>@cyyever , could you comment on what's the situation you were trying to fix ?

**Comment by cyyever** [comment]
> @cyyever any comments ?
> [#17598 (comment)](https://github.com/opencv/opencv/issues/17598#issuecomment-647125305)
> 
> > @cyyever , could you comment on what's the situation you were trying to fix ?

My default g++ is g++-10, without this patch, nvcc would use g++-10 and because no CUDA supports g++-10, it would failed to detect my CUDA arch even if I specify CUDA_HOST_COMPILER.

**Comment by tomoaki0705** [comment]
@cyyever, @NHarishGit 
I appreciate if you could have a look on #17671 
If my understanding is correct, this patch should satisfy all three of us.

**Comment by NHarishGit** [comment]
> @cyyever, @NHarishGit
> I appreciate if you could have a look on #17671
> If my understanding is correct, this patch should satisfy all three of us.

Looks good to me @tomoaki0705. Do you know tentative plan for 4.4.0 release ?

**Comment by tomoaki0705** [comment]
@NHarishGit , thank you for checking.  I don't know the release schedule of 4.4.0. Best

**Pull request merged**